### PR TITLE
Ignore tun/tap network interfaces when calculating network total recieved and sent

### PR DESCRIPTION
**Pull request by relkochta**
TUN/TAP devices are a feature of the Linux kernel which allow userspace applications to create virtual network interfaces. As such, they are not physical network devices and should not be used while calculating the total network usage since the system's boot.

This allows proper calculations for users of VPNs in particular.

### CODE REVIEW

File: src/load-graph.cpp [hunk 1]
```diff
@@ -555,6 +555,28 @@ get_net (LoadGraph *graph)
             and not (netload.flags & (1 << GLIBTOP_NETLOAD_ADDRESS)))
             continue;
 
+        /* Skip tun and tap interfaces */
+#ifdef __linux__
```
Comment (Lightn1ngStr1ke) [inline src/load-graph.cpp]: By doing this, tun/tap devices that have been renamed will be properly ignored.

**Comment by Lightn1ngStr1ke** [comment]
Umm how do I undo the last commit lol

**Comment by Lightn1ngStr1ke** [comment]
undone lol - sorry about all this, I'm super noob when it comes to git

i was doing some code timing earlier and forgot to remove it before commiting

File: src/load-graph.cpp [hunk 2]
```diff
@@ -555,6 +555,25 @@ get_net (LoadGraph *graph)
...
+        char path[strlen(ifnames[i]) + 26];
+        sprintf(path, "/sys/class/net/%s/tun_flags", ifnames[i]);
+        if (!access(path, F_OK))
+            continue;
```
Comment (sc0w) [inline src/load-graph.cpp]: 'continue' without loop?

File: src/load-graph.cpp [hunk 3]
```diff
@@ -555,6 +555,25 @@ get_net (LoadGraph *graph)
...
+#else
+        int ifnamelen = strlen(ifnames[i]);
+        if (!strncmp(ifnames[i], "tun", 3) || !strncmp(ifnames[i], "tap", 3))
+        {
+            bool cont = false;
+            for (int j = 3; j < ifnamelen; ++j)
+                if (!isdigit((int)((ifnames[i])[j])))
+                    cont = true;
+            if (!cont)
+                continue;
```
Comment (sc0w) [inline src/load-graph.cpp]: the same here, 'continue' without loop

**Comment by bdejean** [comment]
Hi, would you like be to directly add this in libgtop2 ?

**Comment by Lightn1ngStr1ke** [comment]
@bdejean how exactly would that be accomplished?

**Comment by bdejean** [comment]
I think that tap devices have an ARP, so checking if_flags for NOARP won't work.
Could you please confirm that a `if_flags & (1L << GLIBTOP_IF_FLAGS_NOARP)` is false for a tap device ?

So I could add new bit for if_flags, either call it ISATAP (like Linux does) or ISVIRTUAL in the next libgtop release.

**Comment by Lightn1ngStr1ke** [comment]
@bdejean Hmm, so it seems that tap devices do not have an ARP and it's easy to figure it out this way. Thank you

**Comment by raveit65** [comment]
I was able to set up a openvpn server and client with a working VPN connection and i can confirm that sending a file with `scp` command from client to server cause a network-load in system-monitor.
I will test the PR tomorow.

**Comment by raveit65** [comment]
Hmm, i see the vpn network load with the net-graph and as `Total received` in m-s-m (vp-server) with your PR.
I used that command  to copy a large file from vpn-client to my vpn-server (10.8.0.1).
`scp Win10_1709_German_x64.iso rave@10.8.0.1:/home/rave`
I use a routed site-to-site config for the vpn-server.
Do i miss something?

**Comment by Lightn1ngStr1ke** [comment]
@raveit65

The idea is to not double-count the usage - i.e. once when it goes through the VPN then again once it actually goes through the physical NIC.

**Comment by raveit65** [comment]
@Lightn1ngStr1ke 
good idea, but how can this be tested?
before --> after

**Comment by Lightn1ngStr1ke** [comment]
@raveit65 What VPN client are you using and what does ifconfig/ip addr show

**Comment by raveit65** [comment]
ip addr from client (fedora 28, opnvpn)

7: tun0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 100
    link/none 
    inet 10.8.0.6 peer 10.8.0.5/32 scope global tun0
       valid_lft forever preferred_lft forever
    inet6 fe80::39dc:2504:720c:df98/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever

**Comment by Lightn1ngStr1ke** [comment]
@raveit65 let me take a look, that shouldn't happen

**Comment by raveit65** [comment]
@Lightn1ngStr1ke 
What kind of test case did you run to be sure that your changes are reducing the displayed network-load with an openvpn connection?

**Comment by raveit65** [comment]
Can this be closed?

**Comment by relkochta** [comment]
@raveit65 Unless we can figure out why the patch didn't reproducibly work, then I suppose so.

**Comment by lukefromdc** [comment]
This is a BAD idea, as the tun/tap interfaces are used for USB tethering over ADB to defeat cellular controls on tethering. Azilink on the phone creates a local OpenVPN server (unencrypted) accessed over the USB cable by Android's ADB (debugging) interface. Carriers cannot easily distinguish Azilink data from any other and if they tried doing so by app name a trivial renaming would defeat them.

**Pull request closed without merging**